### PR TITLE
Support ORDER BY in compiled runtime

### PR DESCRIPTION
**Pull request by henriknyman**
Support ORDER BY of basic types

Implements full `Sort`, but not yet `TopN` or `Top1`, i.e. not optimized for combination with LIMIT yet.

- Introduces notion of tuples for storage of intermediate rows used in joins, aggregations and sorts.
- Adds a new tuple type that implements `Comparable` for use in sorting.
- Generates specialized comparators for primitive Cypher types integers, floats and booleans
- Support for collections of primitive types
- Support for unconditional blocks in code generator module

Full support for the Cypher orderability CIP is upcoming.

### CODE REVIEW

File: community/cypher/acceptance-spec-suite/src/test/resources/cypher/features/OrderByAcceptance.feature [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+Feature: OrderByAcceptance
+
+  Background:
+    Given any graph
```
Comment (Mats-SX) [inline community/cypher/acceptance-spec-suite/src/test/resources/cypher/features/OrderByAcceptance.feature]: This is the loosest possible assumption. Typically it isn't safe to use whenever there is a `MATCH` in the `When` query -- however it is implemented to just use an empty graph (I should change that).

I think it needs to be `Given an empty graph` here.

File: community/cypher/acceptance-spec-suite/src/test/resources/cypher/features/OrderByAcceptance.feature [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      CREATE (:L {a: 3, b: "a"}),
+             (:L {a: 1, b: "b"}),
+             (:L {a: 3, b: "c"}),
+             (:L {a: 4, b: "d"}),
+             (:L {a: 2, b: "e"})
+      """
+    When executing query:
+      """
+      MATCH (n:L)
+      WITH n.a as a, n.b as b
```
Comment (Mats-SX) [inline community/cypher/acceptance-spec-suite/src/test/resources/cypher/features/OrderByAcceptance.feature]: capitalize `as`

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/CodeGenContext.scala [hunk 3]
```diff
@@ -32,18 +32,39 @@ case class Variable(name: String, codeGenType: CodeGenType, nullable: Boolean =
 class CodeGenContext(val semanticTable: SemanticTable, idMap: Map[LogicalPlan, Id], val namer: Namer = Namer()) {
 
   private val variables: mutable.Map[String, Variable] = mutable.Map()
+  private val projectedVariables: mutable.Map[String, Variable] = mutable.Map.empty
   private val probeTables: mutable.Map[CodeGenPlan, JoinData] = mutable.Map()
   private val parents: mutable.Stack[CodeGenPlan] = mutable.Stack()
   val operatorIds: mutable.Map[Id, String] = mutable.Map()
 
   def addVariable(queryVariable: String, variable: Variable) {
+    //assert(!variables.isDefinedAt(queryVariable)) // TODO: Make the cases where overwriting the value is ok explicit (by using updateVariable)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/CodeGenContext.scala:41]: Is this intentionally left in?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/LogicalPlanConverter.scala [hunk 4]
```diff
@@ -564,20 +567,71 @@ object LogicalPlanConverter {
...
+      (methodHandle, WhileLoop(variable, loopDataGenerator, actions) :: tl)
+    }
+  }
+
+  private def sortAsCodeGenPlan(sort: plans.Sort) = new CodeGenPlan with SingleChildPlan {
+
+    override val logicalPlan = sort
+
+    override def consume(context: CodeGenContext, child: CodeGenPlan) = {
+      val variablesToKeep = context.getProjectedVariables // TODO: Intersect/replace with usedVariables(innerBlock)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/LogicalPlanConverter.scala:606]: Is this left intentionally?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/GetSortedResult.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  override def body[E](generator: MethodStructure[E])(implicit context: CodeGenContext) =
+    generator.trace(opName, Some(this.getClass.getSimpleName)) { l1 =>
+      val variablesToGetFromFields = sortTableInfo.outgoingVariableNameToVariableInfo.collect {
+        case (_, FieldAndVariableInfo(fieldName, queryVariableName, incoming, outgoing))
+          if variablesToKeep.isDefinedAt(queryVariableName) => (outgoing.name, fieldName)
+      }
+      l1.sortTableIterate(sortTableInfo.tableName, sortTableInfo.tupleDescriptor, variablesToGetFromFields) { l2 =>
+        l2.incrementRows()
+        action.body(l2)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/GetSortedResult.scala:41]: What are `l1` and `l2` here? Can they get more descriptive names?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/UnwindPrimitiveCollection.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_2.codegen.ir
+
+import org.neo4j.cypher.internal.compiler.v3_2.codegen.ir.expressions._
+import org.neo4j.cypher.internal.compiler.v3_2.codegen.spi.MethodStructure
+import org.neo4j.cypher.internal.compiler.v3_2.codegen.{CodeGenContext, Variable}
+import org.neo4j.cypher.internal.frontend.v3_2.symbols
+
+case class UnwindPrimitiveCollection(opName: String, collection: CodeGenExpression) extends LoopDataGenerator {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/UnwindPrimitiveCollection.scala:27]: I like how you are optimising for this case.

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/expressions/ListLiteral.scala [hunk 7]
```diff
@@ -30,18 +32,29 @@ case class ListLiteral(expressions: Seq[CodeGenExpression]) extends CodeGenExpre
       instruction.init(generator)
     }
 
-  override def generateExpression[E](structure: MethodStructure[E])(implicit context: CodeGenContext) =
-    structure.asList(expressions.map(e => structure.box(e.generateExpression(structure), e.codeGenType)))
+  override def generateExpression[E](structure: MethodStructure[E])(implicit context: CodeGenContext) = {
+    val cType = codeGenType
+    cType match {
+      case CodeGenType(ListType(_), ListReferenceType(innerRepr)) if RepresentationType.isPrimitive(innerRepr) =>
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/expressions/ListLiteral.scala]: `cType` could be inlined, using `case cType@CodeGenType(...`

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/spi/MethodStructure.scala [hunk 8]
```diff
@@ -177,5 +197,27 @@ sealed trait RecordingJoinTableType extends JoinTableType
...
+case object Descending extends SortOrder
+
+case class SortItem(fieldName: String, sortOrder: SortOrder)
+
+/**
+  * What we call tuple here is a fixed length collection of query variable values,
+  * e.g. a result row (or intermediate row).
+  * The TupleDescriptor carries type information used to generate a materialization of this,
+  * e.g. a class with fields.
+  * The order of individual fields are not specified, but is left as an implementation detail.
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/spi/MethodStructure.scala]: are -> is

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledOrderabilityUtils.java [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+            return lhs.compareTo( rhs );
+        }
+    };
+
+    private static Comparator LIST_COMPARATOR = new Comparator<Object>()
+    {
+        @Override
+        public int compare( Object lhs, Object rhs )
+        {
+            Iterator lhsIter = (lhs.getClass().isArray()) ? arrayToIterator( lhs ) : ((List) lhs).iterator();
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledOrderabilityUtils.java:284]: Are we sure that this method will only ever be called with non-null objects that are either arrays or instances of `List`? Perhaps it's worth putting that in a doc comment?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala [hunk 10]
```diff
@@ -75,14 +82,177 @@ class AuxGenerator(val packageName: String, val generator: CodeGenerator) {
     })
   }
 
-  private def newValueTypeName() = {
-    val name = "ValueType" + nameId
+  private def sortResultSign(sortOrder: SortOrder): Int = sortOrder match {
+    case Ascending => -1
+    case Descending => +1
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:87]: Isn't it more convenient to embed these constants with the `SortOrder` objects, and then just ask it directly for its sign?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala [hunk 11]
```diff
@@ -75,14 +82,177 @@ class AuxGenerator(val packageName: String, val generator: CodeGenerator) {
...
+  private def greaterThanSortResult(sortOrder: SortOrder) = -sortResultSign(sortOrder)
+
+  def comparableTypeReference(tupleDescriptor: OrderableTupleDescriptor): TypeReference = {
+    types.getOrElseUpdate(tupleDescriptor,
+      using(generator.generateClass(packageName, newComparableTupleTypeName(),
+      typeRef[Comparable[_]])) { clazz =>
+      tupleDescriptor.structure.foreach {
+        case (fieldName, fieldType: CodeGenType) => clazz.field(lowerType(fieldType), fieldName)
+      }
+      using(clazz.generateMethod(typeRef[Int], "compareTo", param(typeRef[Object], "other"))) { l1 =>
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:100]: What is `l1` here?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala [hunk 12]
```diff
@@ -75,14 +82,177 @@ class AuxGenerator(val packageName: String, val generator: CodeGenerator) {
...
+            sortOrder match {
+              case Ascending =>
+                (lhs, rhs)
+              case Descending =>
+                (rhs, lhs)
+            }
+
+          using(l1.block()) { l2 =>
+            codeGenType match {
+              // TODO: Primitive nodes and relationships including correct ordering of nulls
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:147]: What is a primitive node?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala [hunk 13]
```diff
@@ -75,14 +82,177 @@ class AuxGenerator(val packageName: String, val generator: CodeGenerator) {
...
+                  return -1;
+                } else if (otherValue_a < thisValue_a) {
+                  return +1;
+                }
+                ...
+                */
+                val (thisValueVariable, otherValueVariable) = assignComparatorVariablesFor(l2, reprType)
+
+                using(l2.ifStatement(Expression.lt(thisValueVariable, otherValueVariable))) { l3 =>
+                  l3.returns(Expression.constant(lessThanSortResult(sortOrder)))
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:164]: These `l*` variables, are they convention in the Birk code base? I find them a bit hard to understand.

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/GetSortedResult.scala (hunk 5 above)
Comment (henriknyman) [reply community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/ir/GetSortedResult.scala:41]: They are code generators (`MethodStructure`) at different levels of nesting (of code blocks). `l1`, `l2`, `l3` etc. seemed to be a pre-existing convention when there were multiple levels in the same instruction.

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/CodeGenContext.scala (hunk 3 above)
Comment (henriknyman) [reply community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/CodeGenContext.scala:41]: yes

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledOrderabilityUtils.java (hunk 9 above)
Comment (henriknyman) [reply community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledOrderabilityUtils.java:284]: Yes, these are private comparators and the public entry point to this class (the static `compare` method) handles `null` at the top.
The definition of the `LIST` super type, which uses this comparator is `value instanceof List<?> || value.getClass().isArray()`.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala (hunk 10 above)
Comment (henriknyman) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:87]: Maybe if this was also used somewhere else. `SortOrder` lives within the SPI so it is nice to keep it small and unbothered with implementation details.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala (hunk 12 above)
Comment (henriknyman) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:147]: A node represented by its id as a primitive `long`

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/LogicalPlanConverter.scala (hunk 4 above)
Comment (henriknyman) [reply community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/codegen/LogicalPlanConverter.scala:606]: yes

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala (hunk 13 above)
Comment (henriknyman) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_2/codegen/AuxGenerator.scala:164]: `l1`, `l2`, `l3` etc. seemed to be a pre-existing convention when there were multiple nested levels of generators (`MethodStructure`). You could think of it as blocks in the generated code, and you need to emit code in the right block:
```
{ // l1
  {  // l2
    {  // l3
      // l3.returns()
      return;
    }
  }
}
```

Then again this is not followed everywhere, but at least it is _one_ convention...

**Pull request merged**